src/Libraries/Microsoft.Extensions.AI/Embeddings/NormalizationEmbeddingGenerator.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#if NET // requires generic math interfaces
+using System.Collections.Generic;
+using System.Numerics;
+using System.Numerics.Tensors;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.AI;
+
+/// <summary>A delegating embedding generator that normalizes embedding vectors not already annotated as being normalized.</summary>
+/// <typeparam name="TInput">Specifies the type of the input passed to the generator.</typeparam>
+/// <typeparam name="TEmbeddingValue">Specifies the type of the values in the embedding vectors produced by the generator.</typeparam>
+public sealed class NormalizationEmbeddingGenerator<TInput, TEmbeddingValue> : DelegatingEmbeddingGenerator<TInput, Embedding<TEmbeddingValue>>
+    where TEmbeddingValue : IRootFunctions<TEmbeddingValue>
+{
+    /// <summary>Initializes a new instance of the <see cref="NormalizationEmbeddingGenerator{TInput, TEmbeddingElement}"/> class.</summary>
+    /// <param name="innerGenerator">The underlying <see cref="IEmbeddingGenerator{TInput, Embedding}"/>.</param>
+    public NormalizationEmbeddingGenerator(IEmbeddingGenerator<TInput, Embedding<TEmbeddingValue>> innerGenerator)
+        : base(innerGenerator)
+    {
+    }
+
+    /// <inheritdoc/>
+    public override async Task<GeneratedEmbeddings<Embedding<TEmbeddingValue>>> GenerateAsync(
+        IEnumerable<TInput> values, EmbeddingGenerationOptions? options = null, CancellationToken cancellationToken = default)
+    {
+        var embeddings = await InnerGenerator.GenerateAsync(values, options, cancellationToken).ConfigureAwait(false);
+
+        foreach (var e in embeddings)
+        {
+            if (!e.Normalized)
+            {
+                var normalized = new TEmbeddingValue[e.Vector.Length];
+                TensorPrimitives.Divide(e.Vector.Span, TensorPrimitives.Norm(e.Vector.Span), normalized);
```

Comment:
Should we check if the vector is zero here?

Reply:
I mean the latter, i.e. `[0, 0, ... ,0]` or any other input that could result in a zero norm. I don't think this could ever be returned by a well behaved embedding generator but might make sense for the middleware to be resilient against it.

Reply:
> I mean the latter, i.e. `[0, 0, ..., 0]` or any other input that could result in a zero norm. I don't think this could ever be returned by a well behaved embedding generator but might make sense for the middleware to be resilient against it.

Define resilient? In such a case, this will end up returning a vector of NaNs, which seems like reasonable behavior. What would you recommend instead?

Reply:
My expectation would be to forward the original vector as-is in that case.

Reply:
In other words, I would roughly expect something like this:

```c#
T norm = TensorPrimitives.Norm(e.Vector.Span);
if (T.IsZero(norm) || T.Abs(norm - T.One) < T.CreateTruncating(SomeTolerance))
{
    continue;
}

// Existing normalization code as usual...
```

Reply:
It'd be much more surprising to me if there were such opinions in the implementation.

Reply:
It does tolerate it. It doesn't blow up, throw, fail fast. It produces NaN for garbage input.

Reply:
Technically `[double.Epsilon, -double.Epsilon]` is a vector that defines a polarity but it isn't normalizable because its magnitude evaluates to zero. Overwriting it with NaN seems inappropriate because it loses information even if it is garbage input in the context of embeddings.

Reply:
> Overwriting it with NaN seems inappropriate because it loses information

By definition normalization loses information and the developer is opting into normalization. It'd be surprising if the resulting vectors weren't normalized, so from my perspective allowing through a non-normalized vector would be the bug. Outputting NaNs conveys in the invalidity of the operation, while also ensuring that the likely subsequent operations are appropriately tainted. If someone ends up using CosineSimilarity, for example, with a [double.Epsilon, -double.Epsilon] vector and just does a "use whichever vector has the largest value" to find the best match, they'd end up incorrectly choosing this erroneous vector because it would result in infinity.